src/backend/d3d12/BindGroupD3D12.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BACKEND_D3D12_BINDGROUPD3D12_H_
+#define BACKEND_D3D12_BINDGROUPD3D12_H_
+
+#include "common/BindGroup.h"
+
+#include "d3d12_platform.h"
+
+#include "DescriptorHeapAllocator.h"
+
+namespace backend {
+namespace d3d12 {
+
+    class Device;
+
+    class BindGroup : public BindGroupBase {
+        public:
+            BindGroup(Device* device, BindGroupBuilder* builder);
+
+            void RecordDescriptors(const DescriptorHeapHandle &cbvSrvUavHeapStart, uint32_t* cbvUavSrvHeapOffset, const DescriptorHeapHandle &samplerHeapStart, uint32_t* samplerHeapOffset);
+            uint32_t GetCbvUavSrvDescriptorCount() const;
+            uint32_t GetSamplerDescriptorCount() const;
+            uint32_t GetCbvUavSrvHeapOffset() const;
+            uint32_t GetSamplerHeapOffset() const;
+
+        private:
+            Device* device;
+            uint32_t cbvUavSrvHeapOffset;
+            uint32_t samplerHeapOffset;
+            uint32_t cbvUavSrvCount = 0;
+            uint32_t samplerCount = 0;
+            uint64_t heapSerial = 0;
+
+            union Descriptor {
+                D3D12_CONSTANT_BUFFER_VIEW_DESC cbv;
+                D3D12_UNORDERED_ACCESS_VIEW_DESC uav;
+                D3D12_SHADER_RESOURCE_VIEW_DESC srv;
+                D3D12_SAMPLER_DESC sampler;
+            };
+
+            std::array<Descriptor, kMaxBindingsPerGroup> descriptors;
```

Comment:
Sorry for the conflicting comments, since the SRV UAV CBV and SAMPLER DESC are already precomputed in the views and nxt Sampler, I think storing these offsets would use too much memory for almost no gain. Can we get the DESC's in RecordDescriptors instead?